File: Leetcode/typescript/1497-can-arrange.ts
function canArrange(arr: number[], k: number): boolean {
  const remainderCount: Map<number, number> = new Map();

  for (let num of arr) {
    let remainder = num % k;
    if (remainder < 0) {
      remainder += k;
    }
    remainderCount.set(remainder, (remainderCount.get(remainder) || 0) + 1);
  }

  for (let [rem, count] of remainderCount) {
    if (rem === 0) {
      if (count % 2 !== 0) return false;
    } else if (rem * 2 === k) {
      if (count % 2 !== 0) return false;
    } else {
      const complementCount = remainderCount.get(k - rem) || 0;
      if (count !== complementCount) return false;
    }
  }
  return true;
}

File: Leetcode/python/1497-can-arrange.py
from typing import List

class Solution:
    def canArrange(self, arr: List[int], k: int) -> bool:
        remainderCount = {}

        for num in arr:
            remainder = num % k

            if remainder < 0:
                remainder += k
            if remainder in remainderCount:
                remainderCount[remainder] += 1
            else:
                remainderCount[remainder] = 1

        for rem in remainderCount:
            if rem == 0:
                if remainderCount[rem] % 2 != 0:
                    return False
            elif rem * 2 == k:
                if remainderCount[rem] % 2 != 0:
                    return False
            else:
                if remainderCount[rem] != remainderCount.get(k - rem, 0):
                    return False
        return True
